aten/src/ATen/core/dispatch/DispatchTable.h

```diff
@@ -205,40 +203,36 @@ class DispatchTable final {
     bool is_valid_;
 
     TensorTypeId get_dispatch_key(const Stack* stack, const std::string& operator_name) const {
-      const IValue& first_tensor_arg = torch::jit::peek(
-        *stack,
-        0,
-        reverse_index_of_first_tensor_arg_
-      );
-      // TODO: This will need to get adjusted for multiple dispatch
-      if (C10_UNLIKELY(first_tensor_arg_is_tensor_list_)) {
-        auto tensor_list = first_tensor_arg.toTensorListRef();
-        if (tensor_list.size() == 0) {
-          throw std::runtime_error("Tried to dispatch operator " + operator_name + " based on an empty tensor list. When the first tensor argument of an operator is a tensor list, then it must not be empty.");
+
+      TensorTypeSet ts;
+      for (const auto& ivalue : torch::jit::last(*stack, num_args_)) {
+        if (ivalue.isTensor()) {
+          ts = ts | ivalue.toTensor().type_set();
+        } else if (ivalue.isTensorList()) {
+          for (const auto& tensor : ivalue.toTensorListRef()) {
+            ts = ts | tensor.type_set();
+          }
         }
-        // TODO: Don't use legacy extractor; blocked on c10 understanding
-        // variable
-        return c10::legacyExtractTypeId(tensor_list[0].type_set());
-      } else {
-        return c10::legacyExtractTypeId(first_tensor_arg.unsafeToTensorImpl()->type_set());
       }
+      return at::impl::dispatchTypeId(ts);
     }
   };
 
   static DispatchStrategy get_dispatch_strategy_(const FunctionSchema& schema) {
+    bool is_valid = false;
```

Comment:
nit: This code can become shorter by removing `is_valid` and just returning from the loop. Up to you though, the current implementation is also ok.